cupy/fft/fft.py

```diff
@@ -97,16 +98,29 @@ def _exec_fft(a, direction, value_type, norm, axis, overwrite_x,
 
     batch = a.size // n
 
+    # plan search precedence:
+    # 1. plan passed in as an argument
+    # 2. plan as context manager
+    # 3. cached plan
+    # 4. create a new one
     curr_plan = cufft.get_current_plan()
     if curr_plan is not None:
         if plan is None:
             plan = curr_plan
         else:
             raise RuntimeError('Use the cuFFT plan either as a context manager'
                                ' or as an argument.')
+
     if plan is None:
         devices = None if not config.use_multi_gpus else config._devices
-        plan = cufft.Plan1d(out_size, fft_type, batch, devices=devices)
+        # TODO(leofang): do we need to add the current stream to keys?
```

Comment:
It looks like currently, the plans are associated with the default stream on creation, but when calling the FFT with a plan provided, the plan will be associated wit the current stream. 

https://github.com/cupy/cupy/blob/97573d737beefb7a9ae9f81c294a005ed475bf4f/cupy/cuda/cufft.pyx#L782-L783

In that case, it seems like we shouldn't need to cache the stream as the plan will be reassociated with whatever stream is currently active. Is there a scenario where one wants to do FFTs from separate streams on a single device? If not, then I don't think we would need to redesign anything.

Reply:
> Is there a scenario where one wants to do FFTs from separate streams on a single device?

I feel this is an uncharted territory as running the same FFT plan on multiple streams might drive the plan's internal state and work area nuts. The cuFFT documentation doesn't say anything about this, only that overlapping plan executions with data transfer is possible. I don't even know how we can set up a "stream mutex" to prevent the same plan from being executed on multiple streams, and I'd rather put this burden to end users...

Reply:
> running the same FFT plan on multiple streams

I am not suggesting one would ever use the same plan simultaneously from separate streams, as I don't see how that could possibly work. It seems like a user could potentially try and do that though with the current design, but agree that it is probably a rare scenario and we don't need to try and guard against it here.  

I was asking more if there is a usage scenario for adding the stream attribute on plan creation (and thus also to the plan cache keys), so that multiple plans that are identical aside from stream could be used simultaneously? I agree that we don't need to add this now even if that is a plausible scenario.

Reply:
Hi Gregory, yes I think we are talking about different scenarios:
1. Same transform, **multiple** plan handles, running in parallel on multiple streams (but each handle on a unique stream),
2. Same transform, **same** plan handle, running in parallel on multiple streams,

by "plan handle" I mean `cufftHandle`. 

I agree both scenarios are complex that they should be considered in a separate PR, but just out of curiosity I re-read the doc, and I feel there might be a solution to address **both** scenarios. Quoting [Streamed cuFFT Transforms](https://docs.nvidia.com/cuda/archive/10.2/cufft/index.html#streamed-cufft-transforms):
> Please note that in order to overlap plans using single plan handle user needs to manage work area buffers. Each concurrent plan execution needs it's exclusive work area. Work area can be set by cufftSetWorkArea function.

If I parse this sentence correctly, it is saying that the situation (2) is possible, as long as the plan is executing on distinct pairs of `(stream, work_area)`. What do you think? I feel it's too good to be true, for example when Run 1 is being executed in the midway and we set the work area and stream for the same plan handle for Run 2, wouldn't Run 1 write to the work area of Run 2 and make everything scrambled?

If an internal lock mechanism makes this possible, then in principle the situation (1) is also fixed (by reusing the same handle) as we still need to allocated multiple work areas anyway, although getting distinct handles is easier to manage.

Reply:
There's also a third case here: Same transform, same plan handle, same stream, with multiple FFT launches being able to be overlapped just like any other kernel launch. That's the case that the quoted documentation section is primarily addressing. 

The following common pattern in user code works pretty well:
- create plan, once
- call cufftSetWorkArea, exec FFT N times
- call cudaStreamSynchronize() to wait until all of the FFTs are complete.

I haven't personally tested with setting a separate stream on each invocation, but it is very possible that this would work.

Furthermore, The majority of the memory being used by an allocated plan is the work area that gets allocated by default (when cufftSetAutoAllocation is true). I think it would be a better implementation for the cache if this memory was not allocated at plan creation time -- i.e, always call `cufftSetAutoAllocation(handle, false)` when generating plans, then explicitly assign a work area for each fft() invocation using a normal cupy memory pool as you describe above. The messy part of this is that you do need an event/callback to handle releasing the work area once the FFT completes.

Probably a separate PR to clean this up? But it's pretty important for a cache implementation, with large batch FFTs I've worked on systems where this work area ended up being hundreds of megabytes of GPU memory.

Reply:
Hi @dicta Thanks for your comments. I have a few questions:

> There's also a third case here: Same transform, same plan handle, same stream, with multiple FFT launches being able to be overlapped just like any other kernel launch. 

If multiple such FFTs are launched on the same stream, how can they be overlapped with each other? Did you mean to overlap FFT with other user kernels and/or data transfer? Isn't that already possible / being done?

> The following common pattern in user code works pretty well:
> 
> * create plan, once
> * call cufftSetWorkArea, exec FFT N times
> * call cudaStreamSynchronize() to wait until all of the FFTs are complete.

On the same stream, yes it will, and it is what will be enabled either by the cache. (FYI, it's already possible by explicitly reusing a plan returned by `get_fft_plan()`.) 

On different streams, this is the Scenario 2 I mentioned above, and I feel we will need x unique buffers for x overlapped streams (x <= N).

> I haven't personally tested with setting a separate stream on each invocation, but it is very possible that this would work.

I guess so too.

> Furthermore, The majority of the memory being used by an allocated plan is the work area that gets allocated by default (when cufftSetAutoAllocation is true). I think it would be a better implementation for the cache if this memory was not allocated at plan creation time -- i.e, always call `cufftSetAutoAllocation(handle, false)` when generating plans, then explicitly assign a work area for each fft() invocation using a normal cupy memory pool as you describe above. The messy part of this is that you do need an event/callback to handle releasing the work area once the FFT completes.

This memory pool thing is actually what @grlee77 mentioned in his https://github.com/cupy/cupy/pull/3730#issuecomment-670154850. But I think it's too messy and hard to make it right. You need to have a mutex protecting every unique pairs of `(stream, buffer)` and also a callback. I don't think it's worth the effort. 

Furthermore, CuPy already has its own memory pool, and we should just reuse it instead of layering on top. The lines of code in this simple module will increase significantly, and it's hard to predict/optimize the mempool for every use cases.

Finally, with the recent pains we are experiencing with newer cuFFT (search "fft" in our issues/PRs), I'd rather not count on cuFFT behaving stably predictively, which we'll need for such a mempool. The cuFFT team never provides a strong guarantee for cuFFT's behaviors, nor do they do a good job in documenting the changes where there's one.

> Probably a separate PR to clean this up? 

Clean up what?

> But it's pretty important for a cache implementation, with large batch FFTs I've worked on systems where this work area ended up being hundreds of megabytes of GPU memory.

That's one of the main scenarios I have in mind that a plan cache will be super useful! Another main issue is there's significant slowdown at plan creation time for certain FFTs (for example, #3556).